Add LED functions to cubes-fw

tstana opened the pull request.

Commit:
added led_blink() fn at the start of main to indicate power on status of CUBES board

Commit:
separate directory for LED related fns

tstana commented on utils/led.c:
```diff
@@ -0,0 +1,113 @@
+/*
+ * LEDs masks used to switch on/off LED through GPIOs.
+ */
+#define LEDS_MASK   (uint32_t)0xAAAAAAAA
```
Why are we setting the LEDS_MASK to include all those bits, when we are using only one (GPIO0)?

sonals301 replied:
Have removed it altogether. We won't be needing this `#define` anymore.

tstana replied:
OK!

tstana commented on utils/led.c:
```diff
@@ -0,0 +1,113 @@
+#define LEDS_MASK   (uint32_t)0xAAAAAAAA
+
+
+volatile uint32_t g_gpio_pattern = LEDS_MASK;
```
I'd also make this variable `static`

sonals301 replied:
Yes, okay.

tstana commented on utils/led.c:
```diff
@@ -0,0 +1,113 @@
+ * @brief Interrupt handler for Timer2
+ *
+ */
+void Timer2_IRQHandler(void)
```
The rest of the functions in the `cubes-fw` have the IRQs at the end of the file. Would it be okay to keep to this "convention" here as well?

sonals301 replied:
Yes, sure! Will keep in mind from next time.

tstana commented on utils/led.c:
```diff
@@ -0,0 +1,113 @@
+
+
+volatile uint32_t g_gpio_pattern = LEDS_MASK;
+static uint8_t num_of_blinks, count;
```
I would see the `num_of_blinks` being renamed to a `num_led_toggles` and assigned to `blinks * 2` in the function... but if it makes more sense to you like this, it is perfectly readable as is!

sonals301 replied:
Yes I guess. But if it is okay with you I would like to leave it as is. But your word is final. Let me know.

tstana replied:
Perfectly fine like this!

tstana commented on utils/led.h:
```diff
@@ -0,0 +1,18 @@
+
+void led_blink(led_error_type_t blinks);
+void led_set_high(void);
+
```
I'd also add a `led_turn_off()` or `led_off()` here and implement it in `led.c`.

sonals301 replied:
Sure, done! Implemented `led_turn_off()` in `led.c`

tstana replied:
Still not solved yet, but I will resolve this and send "in-line" comments in the new code.

tstana commented on utils/led.h:
```diff
@@ -0,0 +1,18 @@
+#define UTILS_LED_H_
+
+void led_blink(led_error_type_t blinks);
+void led_set_high(void);
```
I'd rename this to `led_turn_on()` or `led_on()`.

sonals301 replied:
Sure! I changed it to `led_turn_on()`

tstana replied:
Great - thanks!

tstana commented on utils/led_error.h:
```diff
@@ -0,0 +1,18 @@
+/*
+ * led_error.h
```
The error type could be defined inside `led.h` and this file skipped altogether...

sonals301 replied:
Okay, sure!

tstana commented on utils/led_error.h:
```diff
@@ -0,0 +1,18 @@
+typedef enum {
+    LED_RESET = 4,       // Blink 4 times to indicate MSS power-on reset
+    LED_OTHER_ERROR      // Blink 5 times for another type of error
+} led_error_type_t;
```
Would `led_blink_t` perhaps be a better name? And `LED_BLINK_RESET`/`LED_BLINK_OTHER_ERROR`?

sonals301 replied:
Yes boss! okay, got it. Will make the change.

tstana commented on utils/led.c:
```diff
@@ -0,0 +1,113 @@
+        MSS_TIM2_stop();
+
+        /* Set LED high to indicate board's operational status hereafter */
+        led_set_high();
```
Idea: Could we perhaps implement the `HardFault_Handler()` function in `main.c` and have it `led_turn_off()`/`led_off()`, to visually indicate something's wrong?

sonals301 replied:
Yes sure! So you mean in `HardFault_Handler()` we will turn the LED off but for other error we will know the errors through the `LED_BLINK_` type? Is that correct?

tstana replied:
Yes , indeed! But this is to be optionally implemented at a later stage, so please open an issue or add an item in `TODO_CUBES.txt`. I will mark this thread as resolved when you've done one or the other.

sonals301 replied:
Okay! Added a new issue [https://github.com/tstana/cubes-fw/issues/29](url)

tstana commented on utils/led.h:
```diff
@@ -0,0 +1,18 @@
+ * led.h
+ *
+ *  Created on: 29-Jan-2021
+ *      Author: hp
```
hp? 🙂

sonals301 replied:
Lol, sorry! Yeah that is my laptop name. That's how it took this name I guess. I will correct it in the file.

tstana replied:
No worries, it just caught my eye! 🙂

tstana reviewed (commented):
@sonals301 I've opened this PR to review your code. Overall, a very good job; I've added some comments, mostly to help improve the code. Please check and address them.

Commit:
renamed and de-parameterized led_blink() in the main(), replaced it with led_blink_reset()

Commit:
separately defined led_blink() into led_blink_reset() and led_custom_blink()

tstana commented on utils/led.c:
```diff
@@ -27,17 +27,21 @@ volatile uint32_t g_gpio_pattern = LEDS_MASK;
  *
  */
-void led_blink(led_error_type_t blinks)
+void led_blink_reset(void)
```
My apologies if I was unclear in my review yesterday... I actually very much liked the way your function interface looked like for the LED! A single `led_blink()` function with a param. for the number of times the LED should blink is very understandable! I think the new interface is a bit less intuitive.

My suggestion yesterday was to rename the `enum` members to `LED_BLINK_RESET`, `LED_BLINK_...`, not the functions...

sonals301 replied:
No worries! As discussed, `led.c` file will only have two exported functions (in `led.h)` - `led_init()` and `led_custom_blink()`. As for enum fields I will rename them.

tstana replied:
Thanks - but I still think it should be named `led_blink()`, not `custom_blink`...

sonals301 replied:
Yeah. It is fixed now :)

tstana commented on utils/led.c:
```diff
@@ -27,17 +27,21 @@ volatile uint32_t g_gpio_pattern = LEDS_MASK;
 
+/* Local function prototypes */
+void led_set_high(void);
+void led_conf_tim2_periodic(void);
```
I do not see this function as being used outside the LED module. Therefore, I would:

1. Make it `static`
2. Rename it to `tim2_start()`, or `start_led_timer()`

sonals301 replied:
Yes boss! You are right, I added `static` keyword to both local fns and renamed timer function to `led_tim2_start()`. I want `led_` prefix to be there to indicate its origin module

tstana replied:
OK, if you like it that way, it's fine with me. I don't see a point in knowing what the originating module is if the function is static and it cannot be used outside the module, but if you understand it better like this, fine.

Commit:
delete led_error.h, moved content to led.h

Commit:
modify functions and replace with led_init() and led_custom_blink() fns, other minor changes as per Thedi's review comments

sonals301 commented:
Hi Thedi, I have made all the changes and pushed them just now. Please check if it seems okay now. Thanks for reviewing in your vacation time! :)

Commit:
remove commented code

Commit:
working (tested) trial for timer delay function

tstana commented on main.c:
```diff
@@ -64,6 +64,9 @@ int main(void)
 	NVIC_SetPriority(Timer1_IRQn, 1);
 
+	/* For trial */
+	delay_test();
```
I'm guessing this is test code and will be removed - right?

sonals301 replied:
yup

tstana commented on utils/led.c:
```diff
@@ -6,8 +6,8 @@
 #include <stdint.h>
-#include <stdio.h>
-#include <stdlib.h>
+//#include <stdio.h>
```
Please remove dead code, Git should have us covered.

sonals301 replied:
Removed.

tstana commented on utils/led.c:
```diff
@@ -21,8 +21,8 @@ static uint8_t num_of_blinks, count;
 /* Local function prototypes */
-static void led_turn_on(void);
-void led_turn_off(void);
+//static void led_turn_on(void);
+//static void led_turn_off(void);
```
Please remove dead code.

sonals301 replied:
Yeah, removed now after testing.

tstana commented on utils/tim_delay.c:
```diff
@@ -0,0 +1,65 @@
+/*
+ * tim_delay.c
```
how about, simply, `delay.c`?

sonals301 replied:
I have named it `timer_delay.c` if that is okay. If not I will change it to `delay.c` so let me know :)

tstana replied:
Perfectly fine.

tstana commented on utils/tim_delay.c:
```diff
@@ -0,0 +1,65 @@
+#include "drivers/mss_timer/mss_timer.h"
+#include "CMSIS/system_m2sxxx.h"
+#include "tim_delay.h"
+#include "led.h"
```
Can be removed once `delay_test()` is removed. I'm assuming here the plan is to replace the use of MSS_TIM2 in `led` with calls to `delay()`, hopefully I'm right...

sonals301 replied:
Yup, you are absolutely right about that :)

tstana replied:
This line is still present in the latest version of the code. Better to remove it to avoid later confusion.

sonals301 replied:
Yes, I will remove it.

tstana commented on utils/tim_delay.c:
```diff
@@ -0,0 +1,65 @@
+
+
+// Delay counter
+//static volatile uint32_t DelayCounter;
+volatile uint32_t DelayCounter;
```
Did the `static volatile` qualifier not work, or how come it's commented out?

sonals301 replied:
In one of the SO post, it mentioned usually one of the specifiers is to be used and not mix up by using both. Should I use both?

sonals301 replied:
Okay, I made it `static volatile` type.

sonals301 replied:
[https://stackoverflow.com/questions/46010403/static-volatile-vs-static-vs-volatile-in-c](url)
Link for your ref. I do not think it will create any problem.

tstana commented on utils/tim_delay.c:
```diff
@@ -0,0 +1,65 @@
+}
+
+
+void tim_init(void)
```
How about `delay_init_timer()`?

sonals301 replied:
I renamed the file from `tim_delay.c` to `timer_delay.c`. Hence, modified the functions' names to match the file name. `timer_delay_init()` and `timer_delay()`

tstana commented on utils/tim_delay.h:
```diff
@@ -0,0 +1,20 @@
+/*
+ * tim_delay.h
```
See comments in the `.c` file for details on the changes that should trickle over to the `.h`.

tstana commented on utils/tim_delay.h:
```diff
@@ -0,0 +1,20 @@
+#define UTILS_TIM_DELAY_H_
+
+void tim_init(void);
+void delay(uint32_t ms);
```
Just to clarify: Even if you rename this file to `delay.h`, I still think you can keep the `delay()` function named as you have named it. That is, no need to rename it to `delay_delay()`... 🙂

sonals301 replied:
That was just for testing purposes, I removed the function. It is not needed anymore

tstana commented on utils/led.h:
```diff
@@ -20,5 +20,8 @@ typedef enum {
 void led_init(void);
 void led_custom_blink(led_blink_t blinks);
```
`led_blink()`?

sonals301 replied:
ok

tstana commented on utils/led.c:
```diff
@@ -0,0 +1,175 @@
+ * which is equal to 500ms.
+ *
+ */
+void led_tim2_start(void)
```
The function declaration should also be qualified as `static`, same as the prototype.

sonals301 replied:
We are not going to need this function anyway. Because `delay()` fn worked. And i have turned `TIM1` into a 64-bit TIM. So, we do not need function for LED timer.

tstana reviewed (commented):
Thanks for making the changes, @sonals301! Almost there

sonals301 commented:
Hi Thedi, please don't review code just now. I am not done making all the changes yet :) I was testing and running trials to get it working. I just pushed code just in case. 
I will clean it all up after it works and let you know.

Commit:
rename file and fns names and code cleanup

Commit:
code cleanup, fns modified to use timer_delay() instead

Commit:
TIM1->TIM64, add LED blink with 300ms delay in TIM64_IRQ_Handler() to indicate DAQ duration completion, load proper parameters to TIM64 for delay value

tstana commented on msp/msp_exp_handler.c:
```diff
@@ -714,5 +729,8 @@ void Timer1_IRQHandler(void)
 	citiroc_daq_set_hvps_volt(hvps_get_voltage());
 	citiroc_daq_set_hvps_curr(hvps_get_current());
-	MSS_TIM1_clear_irq();
+    MSS_TIM64_clear_irq();
```
Tab here instead of spaces.

sonals301 replied:
yes, done

tstana commented on msp/msp_exp_handler.c:
```diff
@@ -695,8 +703,15 @@ void msp_exprecv_syscommand(unsigned char opcode)
+			temp_value = temp_value >> 32;
+			load_value_u = temp_value;
+			load_value_l = (uint32_t)(timer_load_value & 0xFFFFFFFFULL);
+            MSS_TIM64_load_immediate(load_value_u, load_value_l);
+            MSS_TIM64_start();
```
Tabs here instead of spaces.

sonals301 replied:
yes, done

tstana commented on msp/msp_exp_handler.c:
```diff
@@ -714,5 +729,8 @@ void Timer1_IRQHandler(void)
+    MSS_TIM64_clear_irq();
+
+	/* Indicate TIM duration completion - blink LED with 300ms delay*/
+	led_blink(LED_BLINK_DAQ, 300);
```
Not best-practice to add a delay inside an IRQ handler. Better to remove this call.

sonals301 replied:
I had put it there to only check for daq time. I am removing it. I do not think we require it there for the final code anyway.

tstana commented on main.c:
```diff
@@ -51,22 +51,21 @@ int main(void)
-	MSS_TIM1_load_immediate(0xffffffff);
-	MSS_TIM1_enable_irq();
+	/* Init timer to write HK before DAQ end */
+    MSS_TIM64_init(MSS_TIMER_ONE_SHOT_MODE);
+    MSS_TIM64_enable_irq();
```
Tabs here instead of spaces.

sonals301 replied:
yes, done

tstana commented on utils/led.c:
```diff
@@ -6,24 +6,19 @@
  */
```
Can you please add the GPL header (copy over from other files)?

sonals301 replied:
yes, done.

tstana commented on utils/led.h:
```diff
@@ -11,17 +11,14 @@
 
```
Can you please add the GPL header to the top of the file (copy over from other files)?

sonals301 replied:
yes, done

tstana commented on utils/timer_delay.c:
```diff
@@ -1,5 +1,5 @@
+ * timer_delay.c
  *
  *  Created on: 19-Feb-2021
  *      Author: Sonal Shrivastava
```
Can you please add the GPL header to the top of the file (copy over from other files)?

sonals301 replied:
yes, done

tstana commented on utils/timer_delay.h:
```diff
@@ -0,0 +1,15 @@
+ *
+ *  Created on: 19-Feb-2021
+ *      Author: Sonal Shrivastava
+ */
```
Can you please add the GPL header to the top of the file (copy over from other files)?

sonals301 replied:
yes, done

Commit:
addressed review comments, removed delay fn call from IRQ handler, added GPL license text in utils files

sonals301 commented:
Hi Thedi,

Thanks for your review comments. I have addressed them all. Sorry for the late fixes.
Please let me know if there are any other changes required. Thanks!

tstana commented:
Thank you @sonals301 for addressing all comments! Also, very kind of you to have added me as an author to the code, but it is all your work, I was merely complaining about code "quality"... 🙂
